File: prev_exams_and_ans/2010ccc/q5_night_hop/KnightHop.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.util.LinkedList;

public class KnightHop {

	private static LinkedList<State> tries;
	private static int[] finalPos;
	private static boolean[][] isVisitedBefore;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		// Initialize startState, finalPos
		String[] startPosStr = br.readLine().split(" ");
		
		State startState = new State(
			0, 
			new int[] {Integer.parseInt(startPosStr[0]), Integer.parseInt(startPosStr[1])}
		);
		
		String[] finalPosStr = br.readLine().split(" ");
		finalPos = new int[] {Integer.parseInt(finalPosStr[0]), Integer.parseInt(finalPosStr[1])};
		
		// Initialize tries, isVisitedBefore
		tries = new LinkedList<>();
		isVisitedBefore = new boolean[9][9];
		tries.add(startState);
		isVisitedBefore[startState.curPos[0]][startState.curPos[1]] = true;
		
		while (tries.size() > 0) {
			State nextVisit = tries.removeFirst();
			if (isFinished(nextVisit)) {
				System.out.println(nextVisit.level);
				return;
			}
			
			tries.addAll(nexts(nextVisit));
		}
	}
	
	private static LinkedList<State> nexts(State state) {
		LinkedList<State> nextVisits = new LinkedList<>();
		
		int[][] moves = new int[][] {{1, 2}, {2, 1}, {2, -1}, {1, -2}, {-1, -2}, {-2, -1}, {-2, 1}, {-1, 2}};

		for (int[] move: moves) {
			State nextState = new State(
				state.level + 1, 
				new int[] {state.curPos[0] + move[0], state.curPos[1] + move[1]}
			);
			
			if (isInsideBound(nextState) && !isVisitedBefore[nextState.curPos[0]][nextState.curPos[1]]) {
				nextVisits.add(nextState);
				isVisitedBefore[nextState.curPos[0]][nextState.curPos[1]] = true;
			}
		}
		
		return nextVisits;
		
	}
	
	private static boolean isFinished(State state) {
		return state.curPos[0] == finalPos[0] && state.curPos[1] == finalPos[1];
	}
	
	private static boolean isInsideBound(State state) {
		return state.curPos[0] <= 8 && state.curPos[0] >= 1 && state.curPos[1] <= 8 && state.curPos[1] >= 1;
	}

}

class State {
	public int level;
	public int[] curPos;
	
	public State(int inputLevel, int[] inputPos) {
		this.level = inputLevel;
		this.curPos = inputPos;
	}
}

File: prev_exams_and_ans/2010ccc/q5_night_hop/knight_hop.py
def finished(state):
    return state[1] == final_pos

def inside_bound(state):
    return state[1][0] <= 8 and state[1][0] >= 1 and state[1][1] <= 8 and state[1][1] >= 1

def next(state):
    for move in moves:
        next_state = [state[0] + 1, [state[1][0] + move[0], state[1][1] + move[1]]]
        if inside_bound(next_state) and not is_visited_before[next_state[1][0]][next_state[1][1]]:
            tries.append(next_state)
            is_visited_before[next_state[1][0]][next_state[1][1]] = True


from collections import deque

start_pos = [int(x) for x in input().split()]
start_state = [0, start_pos]
final_pos = [int(x) for x in input().split()]

# Initilize tries, is_visited_before
tries = deque()
is_visited_before = [[False for _ in range(9)] for _ in range(9)]
tries.append(start_state)
is_visited_before[start_state[1][0]][start_state[1][1]] = True

moves = [[1, 2], [2, 1], [2, -1], [1, -2], [-1, -2], [-2, -1], [-2, 1], [-1, 2]]

while (len(tries) > 0):
    next_state = tries.popleft()
    if finished(next_state):
        print(next_state[0])
        break
    next(next_state)
